### PR TITLE
Fix unification of recursive typedefs

### LINKED ISSUE #5785: Compiler hanging
Compiler version from nightlies: 
haxe_2016-11-01_development_d32dd16

This code hangs the compiler forever (Js target at least).

```haxe
typedef Observable<T> = {
	function flatten<U>():Stream<U>;
};
typedef Stream<T> = {
	>Observable<T>,
	function takeUntilBy<U>(otherObs:Observable<U>):Stream<U>;
	function bufferWhileBy():Stream<Array<T>>;
};

class Main {

  public static function main() {
    var mouseMoves : Stream<Int> = null;
    mouseMoves.takeUntilBy(mouseMoves);
  } 
}
```

What is happening:

Compiler hangs forever..

What should happen:

Compiler completes.

Note:

changing the code to this make the compiler not to block:

```haxe
typedef Observable<T> = {
	function flatten<U>():Stream<U>;
};
typedef Stream<T> = {
	>Observable<T>,
	function takeUntilBy<U>(otherObs:Stream<U>):Stream<U>;
	function bufferWhileBy():Stream<Array<T>>;
};

class Main {

  public static function main() {
    var mouseMoves : Stream<Int> = null;
    mouseMoves.takeUntilBy(mouseMoves);
  } 
}
```

### LINKED ISSUE #7733: Compiler crash with recursive typedef and class
```haxe
class Main{
	public static function main(){}
}

typedef Recursive<T1> = {
	public function rec<T2>(a:Recursive<T1>):Recursive<T2>;
}

class RecClass<T1> {
	public function rec<T2>(a:Recursive<T1>):Recursive<T2> return a;
}
```
```
$ OCAMLRUNPARAM=b haxe -cp src -main Main

Fatal error: exception Stack overflow
Raised by primitive operation at file "pMap.ml", line 132, characters 38-43
Called from file "_build/src/core/type.ml", line 2220, characters 2-1020
Called from file "_build/src/core/type.ml", line 1776, characters 11-17
Re-raised at file "_build/src/core/type.ml", line 1785, characters 4-11
Called from file "_build/src/core/type.ml", line 1776, characters 11-17
<...>
Called from file "_build/src/core/type.ml", line 1776, characters 11-17
Re-raised at file "_build/src/core/type.ml", line 1785, characters 4-11
Called from file "_build/src/core/type.ml", line 2026, characters 10-21
Called from file "_build/src/core/type.ml", line 2230, characters 4-43
Called from file "pMap.ml", line 132, characters 38-43
Called from file "_build/src/core/type.ml", line 2220, characters 2-1020
```
Latest development

### PR DESCRIPTION
**Pull request by RealyUniqueName**
Fixes #5785 
Fixes #7733

### CODE REVIEW

**Comment by ncannasse** [comment]
I can't confirm this is a 100% working and not introducing any type hole, because it's a separate stack that can be triggered at different steps of unification.

But it seems that the stack should be restored after calling success, or else it will leak.

**Comment by RealyUniqueName** [comment]
I don't see how it would leak unless someone wants to manualy provide a value for `?stack` argument . 
The stack is restored to a previous state upon leaving `try_applay_params_rec`

**Comment by RealyUniqueName** [comment]
And since it uses physical equality to compare a target `Type.t` for apply_params and physical equality for type params list it will trigger only for  when we try to substitute type params in the same `Type.t` with the type params list produced by `pply_params` in a previous iteration. So, I don't think it can introduce a type hole.

**Comment by ncannasse** [comment]
> The stack is restored to a previous state upon leaving `try_applay_params_rec`

Only in case of exception. You want to restore it in the "normal" case as well

**Comment by RealyUniqueName** [comment]
Oh, indeed. Missed that.

**Comment by RealyUniqueName** [comment]
Fixed

**Comment by Simn** [comment]
I struggle to understand why this works. In particular, how does the `old_applied_params == params` check ever hold? I don't see how the physical equality of a type list could ever be the right thing to check.

I mean, I acknowledge that it does fix the specific cases reported and I don't think it would lead to false positives, so I'm fine with merging. However, we should add an explanation comment in the code because this is the kind of stuff that nobody understands when looking at the code in 2 years.

**Comment by RealyUniqueName** [comment]
I've added a comment with explanation. Is it clear?

File: src/core/type.ml [hunk 1]
```diff
@@ -687,7 +687,41 @@ let apply_params ?stack cparams params t =
...
+							Notice the second argument is actually the result of a previous `apply_params` call.
+							And the result of the current call is `Rec< Array<Array<Int>> >`.
+
+							The third call would be:
+								`apply_params Array<T> Array<Array<Int>> Rec<Array<T>>`
+							and so on.
+
+							To stop infinite params application we need to check that we are trying to apply params
+							produced by the previous `apply_params Array<Int> _ Rec<Array<T>>` to the same `Rec<Array<T>>`
+						*)
```
Comment (Aurel300) [inline src/core/type.ml:718]: That's the sort of documentation we need ^^

**Comment by Simn** [comment]
I still don't get why the list equality check holds. I get that the lists contain the same types, but the fact that they _are_ the same physical lists seems like a bit of an assumption. However, I'm not sure how to demonstrate this logic breaking, so as I said I'm fine with merging.

File: src/core/type.ml (hunk 1 above)
Comment (nadako) [reply src/core/type.ml:718]: But is it the sort we deserve? ^^

**Pull request merged**